### PR TITLE
Ensure videos are allocated into all specified splits

### PR DESCRIPTION
**Pull request by ejm714**
This PR fixes the issue that caused the following error:

```
RuntimeError: Early stopping conditioned on metric val_macro_f1 which is not available. Pass in or modify your EarlyStopping callback to use any of the following: train_loss
```

The issue has to do with the `split` column. When we don't provide it in the labels, we use a random split. But we don't guarantee that we have all of the values from the split_proportions dict (e.g. `train` and `val` ) in the split column. With a small number of videos and the default proportions, all videos getting assigned to train (see in `splits.csv`). Since there are no val videos, there is no `val_f1_score`.

I considered implementing this without the random draw and instead calculating the number of samples that need to go into each split based on the proportions provided. However, this gets tricky to ensure based on rounding. I instead went with the while loop that tries different seeds until it finds a draw that works. If someone specifies split proportions of `train: 100, val: 1`, and only provides say 5 videos, there is risk of an infinite loop. I'm not sure how important it is to catch such a situation since 1) we don't expect people to be training on so few videos and 2) it feels hard to accidentally end up in such a situation.

**Bonus fix**: fix bug in ModelCheckpoint which lets monitor be None if there is no early stopping. `mode` must either be `min` or `max`. I set it to `min` to match PTL's default but it doesn't get used if monitor is None.

Plus cleanup to set defaults in `DummyTrainConfig` to avoid repeated code.

### CODE REVIEW

File: zamba/models/config.py [hunk 1]
```diff
@@ -495,19 +496,28 @@ def preprocess_labels(cls, values):
...
-                    k=len(labels),
+                logger.info(
+                    "No 'site' column found so videos will be randomly allocated using split proportions."
                 )
 
+                expected_labels = [k for k, v in values["split_proportions"].items() if v > 0]
+                labels["split"] = ""
+                seed = SPLIT_SEED
+
+                while len(np.setdiff1d(expected_labels, labels.split.unique())):
```
Comment (pjbull) [inline zamba/models/config.py:508]: This has the (improbable) possibility of running forever. Does it make more sense to group by the labels and then assign each video in that group randomly to a split, warning when the group is less than three items?

**Review by pjbull — CHANGES_REQUESTED** [review]

File: zamba/models/config.py [hunk 2]
```diff
@@ -495,19 +496,28 @@ def preprocess_labels(cls, values):
...
-                random.seed(SPLIT_SEED)
-                labels["split"] = random.choices(
-                    list(values["split_proportions"].keys()),
-                    weights=list(values["split_proportions"].values()),
-                    k=len(labels),
+                logger.info(
+                    "No 'site' column found so videos will be randomly allocated using split proportions."
                 )
 
+                expected_labels = [k for k, v in values["split_proportions"].items() if v > 0]
```
Comment (pjbull) [inline zamba/models/config.py:504]: This var name is a little confusing since `labels` means something specific in our context. Can we change?

Comment (ejm714) [reply zamba/models/config.py:504]: totes, will use `expected_splits`

File: zamba/models/config.py (hunk 1 above)
Comment (pjbull) [reply zamba/models/config.py:508]: Just for clarity, I do mean the _labels_ that the videos have here (not "train", "val" and "holdout")

Comment (ejm714) [reply zamba/models/config.py:508]: this can get tricky too. say i have 5 examples of 20 different species (100 videos) and i set my split proportions to be 8/1/1. if i group by species and assign to sets based on 8/1/1 probabilities, all will go to train. however, if i do this across the whole labels file (not within species), these proportions will be valid.

Comment (ejm714) [reply zamba/models/config.py:508]: curious what your desired behavior would be given that data setup

Comment (pjbull) [reply zamba/models/config.py:508]: you could heuristically do 1 video in each to start and then random after that

Comment (ejm714) [reply zamba/models/config.py:508]: clever

**Comment by ejm714** [comment]
@pjbull ready for another look. this now has the added benefit of doing split allocation within species which is something we had desired but hadn't yet implemented. this should help ensure better training results (in addition to fixing the bug)

**Comment by ejm714** [comment]
@pjbull ready for reals this time!

File: zamba/models/config.py [hunk 3]
```diff
@@ -495,22 +495,44 @@ def preprocess_labels(cls, values):
...
+
+                for c in labels.filter(regex="species_").columns:
+                    species_df = labels[labels[c] > 0]
+
+                    # within each species, seed splits by putting one video in each set and then allocate videos based on split proportions
+                    labels.loc[species_df.index, "split"] = expected_splits + random.choices(
+                        list(values["split_proportions"].keys()),
+                        weights=list(values["split_proportions"].values()),
+                        k=len(species_df) - len(expected_splits),
+                    )
```
Comment (pjbull) [inline zamba/models/config.py:528]: This has one weird edge case that is likely rare, so just worth filing an issue for:

```
v0.mp4, antelope
v1.mp4, antelope  # v1 assigned test by antelope grouping
v1.mp4, cow       # subsequently v1 assigned train by cow grouping
v2.mp4, antelope
v4.mp4, cow
v5.mp4, cow

# test set is now missing antelope

**Review by pjbull — APPROVED** [review]
